AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSGraphQLSubscriptionOperation.swift

```diff
@@ -161,4 +160,27 @@ final public class AWSGraphQLSubscriptionOperation<R: Decodable>: GraphQLSubscri
         }
     }
 
+    private func onSubscriptionFailure(_ error: Error) {
+        let errorDescription = "Subscription item event failed with error"
+        guard case let ConnectionProviderError.subscription(_, payload) = error,
+              let errors = payload?["errors"] as? AppSyncJSONValue else {
+            dispatch(result: .failure(APIError.operationError(errorDescription, "", error)))
+            finish()
+            return
+        }
+
+        if let graphQLErrors = try? GraphQLResponseDecoder.decodeAppSyncErrors(errors),
+           let extensions = graphQLErrors.first?.extensions,
+           let recoveryMessage = graphQLErrors.first?.message,
+           case let .string(errorTypeValue) = extensions["errorType"] {
+                let appSyncError = AppSyncErrorType(errorTypeValue)
+                dispatch(result: .failure(APIError.operationError(errorDescription, recoveryMessage, appSyncError)))
+                finish()
+                return
+        }
```

Comment:
The field "errorType" is coming back from AppSync services is very AppSync specific and isn't really part of the graphQL spec. I recall [DataStore code](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/ProcessMutationErrorFromCloudOperation.swift#L81-L94) checks inside the "extensions" for "errorType" instead of just "errorType" at the top level. And this was because the [merging](https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/API/AWSAPICategoryPlugin/Support/Utils/GraphQLResponseDecoder+DecodeError.swift#L43) happens in the API plugin when decoding a service response with errors like this:
```
{
 "errors": [
   {
     "message":"",
     "extensions: {}
     "errorType:"someError",
   }
 ]
}
```

will transform it into
```
// [GraphQLError] type
{
 "errors": [
   {
     "message":"", // merging logic skips this
     "extensions: { 
        "errorType": "someError"
     }
   }
 ]
}
```

I think the logic you have here to special case it and return an AppSyncErrorType would work for the purpose for DataStore as the client if all it needs is the errorType, but I think it may be more future proof to return `[GraphQLError]` so for all clients dealing with failed subscription connections can handle it in their own way, giving back as much data as possible. How about `let graphQLResponseError = GraphQLResponseError<R>.error(graphQLErrors)` ([GraphQLResponseError](https://github.com/aws-amplify/amplify-ios/blob/main/Amplify/Categories/API/Response/GraphQLResponse.swift))?
```
private func onSubscriptionFailure<R: Decodable>(_ error: Error, responseType: R.Type) {
        let errorDescription = "Subscription item event failed with error"
        if case let ConnectionProviderError.subscription(_, payload) = error,
              let errors = payload?["errors"] as? AppSyncJSONValue,
              let graphQLErrors = try? GraphQLResponseDecoder.decodeAppSyncErrors(errors) {
            let graphQLResponseError = GraphQLResponseError<R>.error(graphQLErrors)
            dispatch(result: .failure(APIError.operationError(errorDescription, "", graphQLResponseError)))
            finish()
            return
        }
```

Reply:
then on the DataStore client code, you would get the operationError, and `underlyingError as? GraphQLResponseError<MutationSync<AnyModel>>`
https://github.com/aws-amplify/amplify-ios/blob/main/AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/IncomingAsyncSubscriptionEventPublisher.swift#L22

Reply:
thanks! makes perfectly sense, I didn't think about returning a `GraphQLResponseError`